KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

jeqo opened the pull request:
Part of [KIP-821](https://cwiki.apache.org/confluence/display/KAFKA/KIP-821%3A+Connect+Transforms+support+for+nested+structures) implementation.

Follow up from https://github.com/apache/kafka/pull/12637

Sets foundations to support nested fields on SMTs.

Includes the following changes:

- Introduces a new modules inside `o.a.k.connect.transforms`: `fields`
- FieldPath interface and 2 implementations: `SingleFieldPath` and `MultiFieldPaths`
- Utils to hold and update Map, Struct and Schema values.

SMTs will be updated in following PRs to complete the scope of KIP-821.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

jeqo committed:
KAFKA-14226: feat(connect:transform): Adapt ExtractField SMT to use nested fields

jeqo committed:
KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

jeqo commented:
cc @C0urante @yashmayya @mimaison

C0urante commented:
@jeqo This is difficult to review without seeing how this code is actually used. Maybe we could move incrementally and introduce a commit that only touches on 1-3 SMTs, and only introduces the internal changes (i.e., methods in the `SingleFieldPath` and `MultiFieldPaths` classes) necessary in order to touch on those SMTs?

C0urante commented:
I'm also wondering if it's necessary to have the `SingleFieldPath` class at all. Would it be significantly more expensive to just use the `MultiFieldPaths` class for everything for now? It'd reduce the complexity of the implementation and probably make review easier.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,581 @@
+        }
+
+        public void addStep(String step) {
+            if (path != null) path = null;
```
Doesn't this disallow examples like the one discussed on the prior PR [here](https://github.com/apache/kafka/pull/12637#discussion_r1148675187)?

jeqo replied:
Thanks for reminding me about this one! Good catch. 
Agree, the suggested operation won't be possible with the current Trie definition as paths are only kept on the "leafs".
I think as preparation to support this we can remove this condition and keep a paths also on nested nodes, see [a65bcfc](https://github.com/apache/kafka/pull/15379/commits/a65bcfc5a9ae7c2fef96135a7416f2eadc80a783). Though this won't be enough to support this feature: MultiFieldPaths will also need to be adapted on fieldAndValueFrom operation to collect values, not only from "leafs" but from intermediate nodes as well. I guess we can do this once SMTs like HeaderFrom will be introduced?

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,581 @@
+
+        public void addStep(String step) {
+            if (path != null) path = null;
+            steps.put(step, new TrieNode());
```
We could refactor this method to return the `Node` for the new step and do something like this:

```java
return steps.computeIfAbsent(step, ignored -> new Node());
```

We could also use `computeIfAbsent` in `addLeaf`.

Then, `Trie::insert` could be simplified:

```java
    public void insert(SingleFieldPath path) {
        Node current = root;

        for (String step : path.stepsWithoutLast()) {
            current = current.addStep(step);
        }

        final String step = path.lastStep();
        current.addLeaf(step);
    }
```

jeqo replied:
Good one, I like the simplification. Applying this suggestion.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,581 @@
+            steps.put(step, new TrieNode(path));
+        }
+
+        public TrieNode get(String step) {
+            return steps.get(step);
+        }
+
+        public Optional<TrieNode> find(String step) {
+            return Optional.ofNullable(steps.get(step));
+        }
```
Aren't these two methods essentially identical? `get` will return `null` if the step isn't found, `find` will return an empty `Optional`.

IMO we should either get rid of `find` entirely and just do standard null checks for return values from it when calling it, or alter `get` to throw an exception if the step _must_ be present but for some reason isn't.

jeqo replied:
Agree, I agree the Optional variants do not make sense as not exposed to users, but used only by MultiFieldPath. Applying suggestion

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Refactor FieldPath abstractions:
- Rename configDef mutation method
- Simpler MultiFieldPaths constructors
- Ensure ordering on MultiFieldPaths mutations to results values/schemas
- Expose additional state required by SMTs: trie size, and original path
- Trie: align access to Trie methods

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

refactor: simplify TrieNode insertion

Applying suggestion [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1556838282

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

refactor: Remove Optional variants

Applying suggestion [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1556852728

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,517 @@
+        return null;
+    }
+
+    /**
+     * Access {@code Map} fields and apply functions to update field values.
+     *
+     * <p>If path is not found, no function is applied, and the path is ignored.
+     *
+     * <p>Other fields keep values from original struct.
+     *
+     * @param originalValue schema-based data value
+     * @param whenFound     function to apply when current path(s) is/are found
+     * @return updated data value
+     */
+    public Map<String, Object> updateValueFrom(
+        Map<String, Object> originalValue,
+        MapValueUpdater whenFound
+    ) {
+        return updateValue(originalValue, 0, whenFound,
+            (originalParent, updatedParent, fieldPath, fieldName) -> {
+                // filter out
+            },
+            (originalParent, updatedParent, fieldPath, fieldName) ->
+                updatedParent.put(fieldName, originalParent.get(fieldName)));
+    }
```
This isn't used in non-testing code; can we remove it for now and add it back in a follow-up if it's necessary for other SMTs?

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,517 @@
+        return updatedParent;
+    }
+
+    /**
+     * Access {@code Struct} fields and apply functions to update field values.
+     *
+     * <p>If path is not found, no function is applied, and the path is ignored.
+     *
+     * <p>Other fields keep values from original struct.
+     *
+     * @param originalSchema original struct schema
+     * @param originalValue  schema-based data value
+     * @param updatedSchema  updated struct schema
+     * @param whenFound      function to apply when current path(s) is/are found
+     * @return updated data value
+     */
+    public Struct updateValueFrom(
+        Schema originalSchema,
+        Struct originalValue,
+        Schema updatedSchema,
+        StructValueUpdater whenFound
+    ) {
+        return updateValue(originalSchema, originalValue, updatedSchema, 0, whenFound,
+            (originalParent, originalField, updatedParent, updatedField, fieldPath) -> {
+                // filter out
+            },
+            (originalParent, originalField, updatedParent, nullUpdatedField, nullFieldPath) ->
+                updatedParent.put(originalField.name(), originalParent.get(originalField)));
+    }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,517 @@
+                        update.apply(
+                            originalValue,
+                            field,
+                            updated,
+                            updateSchema.field(field.name()),
+                            this
+                        );
+                    } else {
+                        if (field.schema().type() == Type.STRUCT) {
+                            Struct fieldValue = updateValue(
+                                field.schema(),
+                                originalValue.getStruct(field.name()),
+                                updateSchema.field(field.name()).schema(),
+                                step + 1,
+                                update,
+                                notFound,
+                                others
+                            );
+                            updated.put(field.name(), fieldValue);
+                        } else {
+                            // add back to not found and apply others, as only leaf values are updated
+                            found = false;
+                            others.apply(originalValue, field, updated, null, this);
+                        }
+                    }
+                } else {
+                    others.apply(originalValue, field, updated, null, this);
+                }
+            }
+        }
+        if (!found) {
+            notFound.apply(
+                originalValue,
+                null,
+                updated,
+                updateSchema.field(stepAt(step)),
+                this);
+        }
+        return updated;
+    }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,517 @@
+        MapValueUpdater update,
+        MapValueUpdater notFound,
+        MapValueUpdater others
+    ) {
+        if (originalValue == null) return null;
+        Map<String, Object> updatedParent = new HashMap<>(originalValue.size());
+        boolean found = false;
+        for (Map.Entry<String, Object> entry : originalValue.entrySet()) {
+            String fieldName = entry.getKey();
+            Object fieldValue = entry.getValue();
+            if (steps.get(step).equals(fieldName)) {
+                found = true;
+                if (step < lastStepIndex()) {
+                    if (fieldValue instanceof Map) {
+                        Map<String, Object> updatedField = updateValue(
+                            (Map<String, Object>) fieldValue,
+                            step + 1,
+                            update,
+                            notFound,
+                            others);
+                        updatedParent.put(fieldName, updatedField);
+                    } else {
+                        // add back to not found and apply others, as only leaf values are updated
+                        found = false;
+                        others.apply(originalValue, updatedParent, null, fieldName);
+                    }
+                } else {
+                    update.apply(originalValue, updatedParent, this, fieldName);
+                }
+            } else {
+                others.apply(originalValue, updatedParent, null, fieldName);
+            }
+        }
+
+        if (!found) {
+            notFound.apply(originalValue, updatedParent, this, stepAt(step));
+        }
+
+        return updatedParent;
+    }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,517 @@
+        return updated;
+    }
+
+    /**
+     * Prepares a new schema based on an original one, and applies an update function
+     * when the current path(s) is found.
+     *
+     * <p>If path is not found, no function is applied, and the path is ignored.
+     *
+     * <p>Other fields are copied from original schema.
+     *
+     * <p>A copy of the {@code Schema} is used as a base for the updated schema.
+     *
+     * @param originalSchema baseline schema
+     * @param whenFound      function to apply when current path(s) is/are found
+     * @return an updated schema. Resulting schemas are usually cached for further access
+     */
+    public Schema updateSchemaFrom(Schema originalSchema, StructSchemaUpdater whenFound) {
+        SchemaBuilder updated = SchemaUtil.copySchemaBasics(originalSchema, SchemaBuilder.struct());
+        return updateSchema(originalSchema, updated, 0, whenFound,
+            (schemaBuilder, field, fieldPath) -> { /* ignore */ },
+            (schemaBuilder, field, fieldPath) -> schemaBuilder.field(field.name(), field.schema()));
+    }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,517 @@
+            (schemaBuilder, field, fieldPath) -> schemaBuilder.field(field.name(), field.schema()));
+    }
+
+    /**
+     * Prepares a new schema based on an original one, and applies an update function
+     * when the current path(s) is found.
+     *
+     * <p>If path is not found, no function is applied, and the path is ignored.
+     *
+     * <p>Other fields are copied from original schema.
+     *
+     * @param originalSchema        baseline schema
+     * @param baselineSchemaBuilder baseline schema build, if changes to the baseline
+     *                              are required before copying original
+     * @param whenFound             function to apply when current path(s) is/are found.
+     * @return an updated schema. Resulting schemas are usually cached for further access.
+     */
+    public Schema updateSchemaFrom(
+        Schema originalSchema,
+        SchemaBuilder baselineSchemaBuilder,
+        StructSchemaUpdater whenFound
+    ) {
+        return updateSchema(originalSchema, baselineSchemaBuilder, 0, whenFound,
+            (schemaBuilder, field, fieldPath) -> { /* ignore */ },
+            (schemaBuilder, field, fieldPath) -> schemaBuilder.field(field.name(), field.schema()));
+    }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,517 @@
+        int step,
+        StructSchemaUpdater matching,
+        StructSchemaUpdater notFound,
+        StructSchemaUpdater others
+    ) {
+        if (operatingSchema.isOptional()) {
+            builder.optional();
+        }
+        if (operatingSchema.defaultValue() != null) {
+            builder.defaultValue(operatingSchema.defaultValue());
+        }
+        boolean matched = false;
+        for (Field field : operatingSchema.fields()) {
+            if (step < steps.size()) {
+                if (steps.get(step).equals(field.name())) {
+                    matched = true;
+                    if (step == lastStepIndex()) {
+                        matching.apply(builder, field, this);
+                    } else {
+                        Schema fieldSchema = updateSchema(
+                            field.schema(),
+                            SchemaBuilder.struct(),
+                            step + 1,
+                            matching,
+                            notFound,
+                            others);
+                        builder.field(field.name(), fieldSchema);
+                    }
+                } else {
+                    others.apply(builder, field, null);
+                }
+            } else {
+                others.apply(builder, field, null);
+            }
+        }
+        if (!matched) {
+            notFound.apply(builder, null, this);
+        }
+        return builder.build();
+    }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,517 @@
+        return builder.build();
+    }
+
+    public String last() {
+        return lastStep();
+    }
+
+    public boolean isEmpty() {
+        return steps.isEmpty();
+    }
```
Unused.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,517 @@
+        return steps.isEmpty();
+    }
+
+    public String stepAt(int i) {
+        return i < steps.size() ? steps.get(i) : "";
+    }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+        return fieldAndValueMap;
+    }
+
+    /**
+     * Access {@code Map} fields and apply functions to update field values.
+     *
+     * <p>If path is not found, no function is applied, and the path is ignored.
+     *
+     * <p>Other fields keep values from original struct.
+     *
+     * @param originalValue schema-based data value
+     * @param whenFound     function to apply when current path(s) is/are found
+     * @return updated data value
+     */
+    public Map<String, Object> updateValuesFrom(
+        Map<String, Object> originalValue,
+        MapValueUpdater whenFound
+    ) {
+        if (trie.isEmpty()) return originalValue;
+        return updateValues(originalValue, trie.root, whenFound,
+            (originalParent, updatedParent, fieldPath, fieldName) -> {
+                // filter out
+            },
+            (originalParent, updatedParent, fieldPath, fieldName) ->
+                updatedParent.put(fieldName, originalParent.get(fieldName)));
+    }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+    @SuppressWarnings("unchecked")
+    private Map<String, Object> updateValues(
+        Map<String, Object> originalValue,
+        TrieNode trieAt,
+        MapValueUpdater matching,
+        MapValueUpdater notFound,
+        MapValueUpdater others
+    ) {
+        if (originalValue == null) return null;
+        Map<String, Object> updatedValue = new LinkedHashMap<>(originalValue.size());
+        Map<String, TrieNode> notFoundFields = new LinkedHashMap<>(trieAt.steps);
+        for (Map.Entry<String, Object> entry : originalValue.entrySet()) {
+            String fieldName = entry.getKey();
+            Object fieldValue = entry.getValue();
+            if (!trieAt.isEmpty()) {
+                if (trieAt.contains(fieldName)) {
+                    notFoundFields.remove(fieldName);
+                    TrieNode trieValue = trieAt.get(fieldName);
+                    if (trieValue.isLeaf()) {
+                        matching.apply(originalValue, updatedValue, trieValue.path, fieldName);
+                    } else {
+                        if (fieldValue instanceof Map) {
+                            Map<String, Object> updatedField = updateValues(
+                                (Map<String, Object>) fieldValue,
+                                trieValue,
+                                matching, notFound, others);
+                            updatedValue.put(fieldName, updatedField);
+                        } else {
+                            // add back to not found and apply others, as only leaf values are updated
+                            notFoundFields.put(fieldName, trieValue);
+                            others.apply(originalValue, updatedValue, null, fieldName);
+                        }
+                    }
+                } else {
+                    others.apply(originalValue, updatedValue, null, fieldName);
+                }
+            } else {
+                others.apply(originalValue, updatedValue, null, fieldName);
+            }
+        }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+        return updatedValue;
+    }
+
+    /**
+     * Access {@code Struct} fields and apply functions to update field values.
+     *
+     * <p>If path is not found, no function is applied, and the path is ignored.
+     *
+     * <p>Other fields keep values from original struct.
+     *
+     * @param originalSchema original struct schema
+     * @param originalValue  schema-based data value
+     * @param updatedSchema  updated struct schema
+     * @param whenFound      function to apply when current path(s) is/are found
+     * @return updated data value
+     */
+    public Struct updateValuesFrom(
+        Schema originalSchema,
+        Struct originalValue,
+        Schema updatedSchema,
+        StructValueUpdater whenFound
+    ) {
+        if (trie.isEmpty()) return originalValue;
+        return updateValues(originalSchema, originalValue, updatedSchema, trie.root, whenFound,
+            (originalParent, originalField, updatedParent, updatedField, fieldPath) -> {
+                // filter out
+            },
+            (originalParent, originalField, updatedParent, nullUpdatedField, nullFieldPath) ->
+                updatedParent.put(originalField.name(), originalParent.get(originalField)));
+    }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+                                matching, notFound, others
+                            );
+                            updatedValue.put(updateSchema.field(field.name()), fieldValue);
+                        } else {
+                            // add back to not found and apply others, as only leaf values are updated
+                            notFoundFields.put(field.name(), trieValue);
+                            others.apply(originalValue, field, updatedValue, null, null);
+                        }
+                    }
+                } else {
+                    others.apply(originalValue, field, updatedValue, null, null);
+                }
+            } else {
+                others.apply(originalValue, field, updatedValue, null, null);
+            }
+        }
+        for (Map.Entry<String, TrieNode> entry : notFoundFields.entrySet()) {
+            String fieldName = entry.getKey();
+            TrieNode trieValue = entry.getValue();
+            if (trieValue.isLeaf()) {
+                notFound.apply(
+                    originalValue,
+                    null,
+                    updatedValue,
+                    updateSchema.field(fieldName),
+                    trieValue.path
+                );
+            } else {
+                Struct fieldValue = updateValues(
+                    SchemaBuilder.struct().build(),
+                    null,
+                    updateSchema.field(fieldName).schema(),
+                    trieValue,
+                    matching, notFound, others
+                );
+                updatedValue.put(updateSchema.field(fieldName), fieldValue);
+            }
+        }
+        return updatedValue;
+    }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+        return updatedValue;
+    }
+
+    /**
+     * Prepares a new schema based on an original one, and applies an update function
+     * when the current path(s) is found.
+     *
+     * <p>If path is not found, no function is applied, and the path is ignored.
+     *
+     * <p>Other fields are copied from original schema.
+     *
+     * <p>A copy of the {@code Schema} is used as a base for the updated schema.
+     *
+     * @param originalSchema baseline schema
+     * @param whenFound      function to apply when current path(s) is/are found
+     * @return an updated schema. Resulting schemas are usually cached for further access
+     */
+    public Schema updateSchemaFrom(
+        Schema originalSchema,
+        StructSchemaUpdater whenFound
+    ) {
+        if (trie.isEmpty()) return originalSchema;
+        SchemaBuilder updated = SchemaUtil.copySchemaBasics(originalSchema, SchemaBuilder.struct());
+        return updateSchema(originalSchema, updated, trie.root, whenFound,
+            (schemaBuilder, field, fieldPath) -> { /* ignore */ },
+            (schemaBuilder, field, fieldPath) -> schemaBuilder.field(field.name(), field.schema()));
+    }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+    }
+
+
+    /**
+     * Prepares a new schema based on an original one, and applies an update function
+     * when the current path(s) is found.
+     *
+     * <p>If path is not found, no function is applied, and the path is ignored.
+     *
+     * <p>Other fields are copied from original schema.
+     *
+     * @param originalSchema        baseline schema
+     * @param baselineSchemaBuilder baseline schema build, if changes to the baseline
+     *                              are required before copying original
+     * @param whenFound             function to apply when current path(s) is/are found.
+     * @return an updated schema. Resulting schemas are usually cached for further access.
+     */
+    public Schema updateSchemaFrom(
+        Schema originalSchema,
+        SchemaBuilder baselineSchemaBuilder,
+        StructSchemaUpdater whenFound
+    ) {
+        if (trie.isEmpty()) return originalSchema;
+        return updateSchema(originalSchema, baselineSchemaBuilder, trie.root, whenFound,
+            (schemaBuilder, field, fieldPath) -> { /* ignore */ },
+            (schemaBuilder, field, fieldPath) -> schemaBuilder.field(field.name(), field.schema()));
+    }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+                if (!trieAt.contains(field.name())) {
+                    toOtherFields.apply(baseSchemaBuilder, field, null);
+                } else {
+                    notFoundFields.remove(field.name());
+                    TrieNode trieNode = trieAt.get(field.name());
+                    if (trieNode.isLeaf()) {
+                        whenFound.apply(baseSchemaBuilder, field, trieNode.path);
+                    } else {
+                        if (field.schema().type() == Type.STRUCT) {
+                            Schema fieldSchema = updateSchema(
+                                field.schema(),
+                                SchemaBuilder.struct(),
+                                trieNode,
+                                whenFound, whenNotFound, toOtherFields);
+                            baseSchemaBuilder.field(field.name(), fieldSchema);
+                        } else {
+                            toOtherFields.apply(baseSchemaBuilder, field, null);
+                        }
+                    }
+                }
+            } else {
+                toOtherFields.apply(baseSchemaBuilder, field, null);
+            }
+        }
+        for (Map.Entry<String, TrieNode> entry : notFoundFields.entrySet()) {
+            String fieldName = entry.getKey();
+            TrieNode trieValue = entry.getValue();
+            if (trieValue.isLeaf()) {
+                whenNotFound.apply(baseSchemaBuilder, null, trieValue.path);
+            } else {
+                Schema fieldSchema = updateSchema(
+                    SchemaBuilder.struct().build(),
+                    SchemaBuilder.struct(),
+                    trieValue,
+                    whenFound, whenNotFound, toOtherFields);
+                baseSchemaBuilder.field(fieldName, fieldSchema);
+            }
+        }
+        return baseSchemaBuilder.build();
+    }
```
Also unused outside of testing code.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/MultiFieldPaths.java:
```diff
@@ -0,0 +1,578 @@
+    // Invariants:
+    // - Trie values contain either a nested trie or a field path when it is a leaf.
+    // - A trie flattens overlapping paths (e.g. `foo` and `foo.bar` in V2, only `foo.bar` would be kept)
+    static class Trie {
```
Since this class is static, perhaps we could pull it out into its own file?

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Remove unused methods to be introduced in future PRs.

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Move Trie class up

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Remove unused interfaces to be introduced in future PRs.

C0urante commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ValueToKeyTest.java:
```diff
@@ -60,6 +62,30 @@ public void schemaless() {
+
+        final HashMap<String, Integer> expectedKey = new HashMap<>();
+        expectedKey.put("a", 1);
+        expectedKey.put("b.c", 3);
```
Hmmm... I thought that the nesting structure of the value would be matched in the derived key. Does the KIP go into detail anywhere about this? I'm not a huge fan of reusing the dot notation here but I can't remember if this was discussed on the KIP thread or not.

jeqo replied:
Good catch. I don't remember this scenario being discussed in the KIP thread. Was using the original keys as they are meant to be unique, but they could also contain all kind of escape fields making it hard to use. 
Happy to switch into deriving the key schema from the nesting structure. Will make an update to the KIP to make this explicit as well.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,258 @@
+        Struct current = struct;
+        for (String pathSegment : stepsWithoutLast()) {
+            // Check to see if the field actually exists
+            if (current.schema().field(pathSegment) == null) {
```
Should we also check if `current` is null, since even if the field is defined in the schema it might be optional?

(And we should probably add a test case for this.)

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,258 @@
+     */
+    @SuppressWarnings("unchecked")
+    public Object valueFrom(Map<String, Object> map) {
+        if (map == null) return null;
+
+        if (steps.size() == 1) {
+            return map.get(steps.get(0));
+        } else {
+            Map<String, Object> current = map;
+            for (int i = 0; i < steps.size(); i++) {
+                if (current == null) {
+                    return null;
+                }
+                if (i == lastStepIndex()) {
+                    return current.get(steps.get(i));
+                } else {
+                    current = (Map<String, Object>) current.get(steps.get(i));
+                }
+            }
+        }
+        return null;
```
We can clean this up a bit:
```suggestion
        if (map == null) return null;

        Map<String, Object> current = map;
        for (String step : stepsWithoutLast()) {
            current = requireMapOrNull(current.get(step), "nested field access");

            if (current == null)
                return null;
        }
        return current.get(lastStep());
```

Using `requireMapOrNull` (statically imported from `org.apache.kafka.connect.transforms.util.Requirements`) makes this more type-safe, and leveraging `stepsWithoutLast` and `lastStep` helps simplify the logic.

C0urante replied:
Oh, and the `@SuppressWarnings` annotation can be removed if this works.

C0urante commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/field/FieldPathNotationTest.java:
```diff
@@ -0,0 +1,149 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class FieldPathNotationTest {
```
We can simplify these tests with a few utility methods:

```java
private void assertParseV1(String path) {
    assertArrayEquals(
            new String[] {path},
            new SingleFieldPath(path, FieldSyntaxVersion.V1).path());
}

private void assertParseV2(String inputPath, String... expectedSteps) {
    assertArrayEquals(
            expectedSteps,
            new SingleFieldPath(inputPath, FieldSyntaxVersion.V2).path()
    );
}

private void assertParseV2Error(String inputPath, String expectedMessage) {
    ConfigException exception = assertThrows(
            ConfigException.class,
            () -> new SingleFieldPath(inputPath, FieldSyntaxVersion.V2)
    );
    assertEquals(expectedMessage, exception.getMessage());
}
```

Some simplified test cases would look like this:
```java
@Test
void shouldBuildV1WithDotsAndBacktickPair() {
    // Given v1
    // When path contains dots, then single step path
    assertParseV1("foo.bar.baz");
    // When path contains backticks, then single step path
    assertParseV1("foo`bar`");
    // When path contains dots and backticks, then single step path
    assertParseV1("foo.`bar.baz`");
}

@Test
void shouldBuildV2WhenIncludesDotsAndBacktickPair() {
    // Given v2 and fields including dots
    // When backticks are wrapping a field name (i.e. withing edges or between dots)
    // Then build a path with steps separated by dots and not including backticks
    assertParseV2("`foo.bar.baz`", "foo.bar.baz");
    assertParseV2("foo.`bar.baz`", "foo", "bar.baz");
    assertParseV2("`foo.bar`.baz", "foo.bar", "baz");
    assertParseV2("foo.`bar`.baz", "foo", "bar", "baz");
}

@Test
void shouldFailV2WhenIncompleteBackticks() {
    // Given v2
    // When backticks are not closed and not escaped
    // Then it should fail
    assertParseV2Error(
            "`foo.bar.baz",
            "Incomplete backtick pair in path: [`foo.bar.baz], consider adding a backslash before backtick at position 0 to escape it"
    );

    assertParseV2Error(
            "foo.`bar.baz",
            "Incomplete backtick pair in path: [foo.`bar.baz], consider adding a backslash before backtick at position 4 to escape it"
    );

    assertParseV2Error(
            "foo.bar.`baz",
            "Incomplete backtick pair in path: [foo.bar.`baz], consider adding a backslash before backtick at position 8 to escape it"
    );

    assertParseV2Error(
            "foo.bar.`baz\\`",
            "Incomplete backtick pair in path: [foo.bar.`baz\\`], consider adding a backslash before backtick at position 8 to escape it"
    );
}
```

C0urante commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/field/FieldPathNotationTest.java:
```diff
@@ -0,0 +1,149 @@
+        ConfigException e0 = assertThrows(ConfigException.class,
+                () -> new SingleFieldPath("`foo.bar.baz", FieldSyntaxVersion.V2));
+        assertEquals(
+                "Incomplete backtick pair in path: [`foo.bar.baz], consider adding a backslash before backtick at position 0 to escape it",
```
Should we test for the case where a backslash precedes an opening backtick?

jeqo replied:
Good question. In the spec we have: 
> If a backtick is followed by a dot in the field name, then the backtick should be escaped with a backslash to signal that the backtick is part of the name and not closing a backtick pair.

This is missing the scenario were a backtick is included in the field name and it's at the beginning.

I think we should extend this to include "If a bactick it's at the beginning of the field name or is followed by a dot..." and account for this when building the path. wdyt?

C0urante replied:
That makes sense to me 👍

C0urante commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/field/SingleFieldPathTest.java:
```diff
@@ -0,0 +1,67 @@
+
+class SingleFieldPathTest {
+
+    @Test
+    void shouldIncludeEmptyFieldNames() {
+        assertArrayEquals(
+            new String[] {"", "", ""},
+            new SingleFieldPath("..", FieldSyntaxVersion.V2).path()
+        );
+        assertArrayEquals(
+            new String[] {"foo", "", ""},
+            new SingleFieldPath("foo..", FieldSyntaxVersion.V2).path()
+        );
+        assertArrayEquals(
+            new String[] {"", "bar", ""},
+            new SingleFieldPath(".bar.", FieldSyntaxVersion.V2).path()
+        );
+        assertArrayEquals(
+            new String[] {"", "", "baz"},
+            new SingleFieldPath("..baz", FieldSyntaxVersion.V2).path()
+        );
+    }
```
This case can probably be moved to `FieldPathNotationTest`, right?

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Moving tests as suggested [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1571018390

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Refactor valueFrom using Requirements util as suggested [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1570960525

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Support intermediate null valueFrom as suggested [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1570939313

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Refactor notation tests as suggested [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1570998105

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Move Trie-related tests to its right class.

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Remove unused method

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Adding missing tests

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Simplify Trie definition to keep path on nested, not only on leaf nodes

Follow up from suggestion [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1556834149

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Remove MultiFieldPaths. To be added in a following PR.

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1556834149

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,240 @@
+            if (current.schema().field(pathSegment) == null) {
+                return null;
+            }
+            current = current.getStruct(pathSegment);
```
Nit: we could have slightly friendlier error messages here:
```suggestion
            Object subValue = current.get(pathSegment);
            current = requireStructOrNull(subValue, "nested field access");
```

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,240 @@
+        }
+    }
+
+    List<String> stepsWithoutLast() {
```
Nit: can be private
```suggestion
    private List<String> stepsWithoutLast() {
```

C0urante replied:
Also nit: would be nice to move this lower to be next to the other private/should-be-private methods like `lastStep` and `lastStepIndex`.

C0urante commented on connect/transforms/src/main/java/org/apache/kafka/connect/transforms/field/SingleFieldPath.java:
```diff
@@ -0,0 +1,240 @@
+        return steps.toArray(new String[0]);
+    }
+
+    String lastStep() {
```
Nit: can be private
```suggestion
    private String lastStep() {
```

C0urante commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java:
```diff
@@ -111,7 +157,24 @@ public void nonExistentFieldWithSchemaShouldFail() {
             fail("Expected exception wasn't raised");
         } catch (IllegalArgumentException iae) {
-            assertEquals("Unknown field: nonexistent", iae.getMessage());
+            assertEquals("Unknown field: SingleFieldPath{version=V1, path=nonexistent}", iae.getMessage());
```
This error message is less human-readable. Could we preserve the existing one? I don't think the syntax version is necessary; we can just use the path as specified by the user in the transform config.

C0urante commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java:
```diff
@@ -99,6 +133,18 @@ public void nonExistentFieldSchemalessShouldReturnNull() {
     }
 
+    @Test
+    public void nonExistentNestedFieldSchemalessShouldReturnNull() {
```
Don't we need to configure the transform to use field syntax V2?

C0urante commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java:
```diff
@@ -111,7 +157,24 @@ public void nonExistentFieldWithSchemaShouldFail() {
+    }
+
+    @Test
+    public void nonExistentNestedFieldWithSchemaShouldFail() {
```
Don't we need to configure the transform to use field syntax V2?

C0urante commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/ExtractFieldTest.java:
```diff
@@ -111,7 +157,24 @@ public void nonExistentFieldWithSchemaShouldFail() {
+            xform.apply(record);
+            fail("Expected exception wasn't raised");
+        } catch (IllegalArgumentException iae) {
+            assertEquals("Unknown field: SingleFieldPath{version=V1, path=magic.nonexistent}", iae.getMessage());
```
Like with `nonExistentFieldWithSchemaShouldFail`, I don't think the field syntax version needs to be included in the error message. But if it is included, shouldn't it be V2 here instead of V1?

C0urante commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/field/FieldSyntaxVersionTest.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    @ParameterizedTest
+    @CsvSource({"v1,V1", "v2,V2", "V1,V1", "V2,V2"})
+    void shouldGetVersionFromConfig(String input, FieldSyntaxVersion version) {
```
This is pretty slick. Nice!

C0urante commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/field/SingleFieldPathTest.java:
```diff
@@ -0,0 +1,113 @@
+class SingleFieldPathTest {
+
+    @Test void shouldFindField() {
+        SchemaBuilder barSchema = SchemaBuilder.struct().field("bar", Schema.INT32_SCHEMA);
```
Better practice if we operate on fully-built `Schema` instances instead of `SchemaBuilder` objects:
```suggestion
        Schema barSchema = SchemaBuilder.struct().field("bar", Schema.INT32_SCHEMA).build();
```

C0urante replied:
(This applies in several other places in this suite; I've only left one comment to avoid clutter.)

C0urante commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/field/SingleFieldPathTest.java:
```diff
@@ -0,0 +1,113 @@
+        assertNull(new SingleFieldPath("foo.baz.inner", FieldSyntaxVersion.V2).valueFrom(struct));
+    }
+
+    private static SingleFieldPath pathV2(String path) {
+        return new SingleFieldPath(path, FieldSyntaxVersion.V2);
+    }
```
We can simplify things even further by piggybacking off of this with a few extra utility methods:

```java
    private static Field pathV2Field(String path, Schema schema) {
        return pathV2(path).fieldFrom(schema);
    }

    private static Object pathV2Value(String path, Struct struct) {
        return pathV2(path).valueFrom(struct);
    }

    private static Object pathV2Value(String path, Map<String, Object> map) {
        return pathV2(path).valueFrom(map);
    }
```

I think we can replace all direct invocations of `pathV2` with these wrapper methods.

jeqo replied:
Not sure if the proposed methods add much. I'd argue the `pathV2()` util is enough to make the invocations more readable.

C0urante replied:
I found it easier to read/review after doing the refactoring, but I don't consider this a blocker. Can always see how things pan out with future PRs when we add more functionality to the `SingleFieldPath` class.

C0urante commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/field/SingleFieldPathTest.java:
```diff
@@ -0,0 +1,113 @@
+        assertNull(new SingleFieldPath("un.known", FieldSyntaxVersion.V2).valueFrom(map));
+        assertNull(new SingleFieldPath("foo.unknown", FieldSyntaxVersion.V2).valueFrom(map));
+        assertNull(new SingleFieldPath("unknown", FieldSyntaxVersion.V2).valueFrom(map));
+        assertNull(new SingleFieldPath("foo.baz", FieldSyntaxVersion.V2).valueFrom(map));
```
This assertion technically doesn't belong in this case, right? This value does appear in the map, it just happens to be explicitly null. It's also covered above in `shouldFindValueInMap`, so it should be safe to remove from here.

(This same pattern applies in other places in the test suite too.)

C0urante commented on connect/transforms/src/test/java/org/apache/kafka/connect/transforms/field/SingleFieldPathTest.java:
```diff
@@ -0,0 +1,113 @@
+        Map<String, Object> map = new HashMap<>();
+        map.put("foo", foo);
+
+        assertNull(new SingleFieldPath("un.known", FieldSyntaxVersion.V2).valueFrom(map));
```
Don't we want to use `pathV2` (or possibly `pathV2Value`, proposed below) instead of manually instantiating the `SingleFieldPath`?

(This applies to other places in the test suite as well.)

C0urante reviewed (commented):
Thanks @jeqo! This is looking great, I think we're probably one or at most two revisions away from merging.

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Apply suggestion [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1583252760

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Apply suggestion [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1583254337

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Apply suggestion [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1583255231

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Apply suggestion [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1583268273

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Apply suggestion [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1583272352

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Apply suggestion [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1583272495

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Apply suggestion [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1583303335

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Apply suggestion [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1583325144

jeqo commented:
Thanks @C0urante! I've applied most suggestions, PTAL

jeqo committed:
fixup! KAFKA-14226: feat(connect:transform): Introduce FieldPath abstraction

Apply suggestion [1]

[1]: https://github.com/apache/kafka/pull/15379#discussion_r1583304218

C0urante reviewed (approved):
LGTM, thanks for your diligent work on this @jeqo!

Looking forward to the next PR 🚀
